UniversalAlgebra/Subalgebras/Properties.lagda

```diff
@@ -0,0 +1,177 @@
+
+open import Algebras.Basic using (𝓞 ; 𝓥 ; Signature )
+
+module Subalgebras.Properties {𝑆 : Signature 𝓞 𝓥} where
+
+-- imports from Agda and the Agda Standard Library ------------------------------------
+open import Agda.Builtin.Equality using ( _≡_ ; refl )
+open import Agda.Primitive        using ( _⊔_ ; lsuc ; Level ) renaming ( Set to Type )
+open import Data.Product          using ( _,_ ) renaming ( proj₁ to fst ; proj₂ to snd )
+open import Function.Base         using ( _∘_ ; id )
+open import Function.Bundles      using ( Injection )
+open import Relation.Unary        using ( Pred ; _⊆_ )
+import Relation.Binary.PropositionalEquality as PE
+
+-- -- imports from agda-algebras --------------------------------------------------------------
+open import Overture.Preliminaries             using ( ∣_∣ ; ∥_∥ ; _⁻¹ )
+open import Overture.Inverses                  using (  id-is-injective ; ∘-injective ; IsInjective )
+open import Algebras.Basic                     using ( Algebra ; Lift-Alg )
+open import Algebras.Products          {𝑆 = 𝑆} using ( ov )
+open import Homomorphisms.Basic        {𝑆 = 𝑆} using ( ∘-hom ; is-homomorphism ; ∘-is-hom )
+open import Homomorphisms.Isomorphisms {𝑆 = 𝑆} using ( _≅_ ; ≅toInjective ; ≅fromInjective ; ≅-refl
+                                                     ; ≅-sym ; ≅-trans ; Lift-≅ ; mkiso )
+open import Subalgebras.Subalgebras    {𝑆 = 𝑆} using ( _≤_ ; _≥_ ; _IsSubalgebraOfClass_ )
+
+private variable α β γ 𝓧 : Level
+
+
+-- The subalgebra relation is a *preorder*, i.e., a reflexive transitive binary relation.
+
+open _≅_
+≅→≤ : {𝑨 : Algebra α 𝑆}{𝑩 : Algebra β 𝑆} → 𝑨 ≅ 𝑩 → 𝑨 ≤ 𝑩
+≅→≤ φ = (to φ) , ≅toInjective φ
+
+≅→≥ : {𝑨 : Algebra α 𝑆}{𝑩 : Algebra β 𝑆} → 𝑨 ≅ 𝑩 → 𝑨 ≥ 𝑩
+≅→≥ φ = (from φ) , ≅fromInjective φ
+
+
+≤-reflexive : (𝑨 : Algebra α 𝑆) → 𝑨 ≤ 𝑨
+≤-reflexive 𝑨 = (id , λ 𝑓 𝑎 → refl) , Injection.injective id-is-injective
+
+≤-refl : {𝑨 : Algebra α 𝑆} {𝑩 : Algebra β 𝑆} → 𝑨 ≅ 𝑩 → 𝑨 ≤ 𝑩
+≤-refl A≅B = ≅→≤ A≅B
+
+≥-refl : {𝑨 : Algebra α 𝑆}{𝑩 : Algebra β 𝑆} → 𝑨 ≅ 𝑩 → 𝑨 ≥ 𝑩
+≥-refl A≅B = ≅→≤ (≅-sym A≅B)
+
+≤-trans : (𝑨 : Algebra α 𝑆){𝑩 : Algebra β 𝑆}(𝑪 : Algebra γ 𝑆)
+ →        𝑨 ≤ 𝑩 → 𝑩 ≤ 𝑪 → 𝑨 ≤ 𝑪
+
+≤-trans 𝑨 {𝑩} 𝑪 A≤B B≤C = (∘-hom 𝑨 𝑪 ∣ A≤B ∣ ∣ B≤C ∣) , ∘-injective ∥ A≤B ∥ ∥ B≤C ∥
+
+
+≥-trans : (𝑨 : Algebra α 𝑆){𝑩 : Algebra β 𝑆}(𝑪 : Algebra γ 𝑆)
+ →        𝑨 ≥ 𝑩 → 𝑩 ≥ 𝑪 → 𝑨 ≥ 𝑪
+
+≥-trans 𝑨 {𝑩} 𝑪 A≥B B≥C = ≤-trans 𝑪 {𝑩} 𝑨 B≥C A≥B
+
+
+module _ {α β ρ : Level} where
+
+ open import Relation.Binary.Structures {a = (ov α)}{ℓ = (𝓞 ⊔ 𝓥 ⊔ α)} (_≅_ {α}{α})
+
+ open IsPreorder
+ ≤-preorder : IsPreorder _≤_
+ isEquivalence ≤-preorder = record { refl = ≅-refl ; sym = ≅-sym ; trans = ≅-trans }
+ reflexive ≤-preorder = ≤-refl
+ trans ≤-preorder {𝑨}{𝑩}{𝑪} A≤B B≤C = ≤-trans 𝑨 {𝑩} 𝑪 A≤B B≤C
+
+open _≅_
+
+module _ {α β γ : Level}{𝑨 : Algebra α 𝑆}{𝑩 : Algebra β 𝑆}{𝑪 : Algebra γ 𝑆} where
+
+-- If two algebras are isomorphic and one of them is a subalgebra of `𝑨`, then so is the other.
```

Comment:
Ah, good point!  I'll take a look tomorrow.  Thanks!!

Reply:
After looking into this I'm very glad you raised the issue.  It led me to observe the fact that standard library definition of poset is relative to a homogeneous equality... but I've so far been working on the assumption that isomorphism is a heterogeneous equivalence relation.  (I'm even assuming the subalgebra relation itself may be heterogeneous, so a subalgebra can live in a different universe than the algebra to which it "belongs", which I believe is needed for handling the closure operator S on classes of algebras).  I'll open a new issue about this.

Reply:
Heterogeneous preorders, partialorders, etc. with respect to a heterogeneous equivalence will require substantial time and effort to do it right, so let's put that off for now.  Since it's already discussed in #82 , I'll resolve this conversation and merge.